### ISSUE TITLE
[jax2tf] InconclusiveDimensionOperation for scatter op with leading vmapped symbolic dimension

### ISSUE THREAD

**Issue by areiner222** [google/jax#18348]
### Description

Running into an issue when trying to use scatter operations inside a converted batch jax function when gradient is required. Is there a recommended work around if I have an unknown batch dimension?

Repro
```
import jax
from jax.experimental import jax2tf
import tensorflow as tf

def func(x):
    inds = jp.array([0, 2, 4], dtype=jp.int32)
    updates = jp.array([100., 101., 102.], dtype=jp.float32)
    x = x.at[jp.array(inds)].set(updates)
    return x
    
b = 10
x = tf.random.uniform((b, 5))
with tf.GradientTape() as tape:
    tape.watch(x)
    func_convert = jax2tf.convert(jax.vmap(func), polymorphic_shapes=['(b, _)'])
    res = func_convert(x)
    loss = tf.reduce_mean(res)

grads = tape.gradient(loss, x)
```

Results in error
```
File ~/mambaforge/envs//lib/python3.10/site-packages/jax/experimental/export/export.py:850, in _get_vjp_fun.<locals>.fun_vjp_jax(*args_and_out_cts_flat_jax)
    846   return res_flat
    848 args_flat_jax, out_cts_flat_jax = util.split_list(args_and_out_cts_flat_jax,
    849                                                   [len(in_avals)])
--> 850 _, pullback_jax = jax.vjp(flattened_primal_fun_jax, *args_flat_jax)
    851 return pullback_jax(out_cts_flat_jax)

    [... skipping hidden 7 frame]

File ~/mambaforge/envs//lib/python3.10/site-packages/jax/experimental/export/export.py:844, in _get_vjp_fun.<locals>.fun_vjp_jax.<locals>.flattened_primal_fun_jax(*args_flat)
    842 def flattened_primal_fun_jax(*args_flat):
    843   args, kwargs = in_tree.unflatten(args_flat)
--> 844   res = primal_fun(*args, **kwargs)
    845   res_flat, _ = tree_util.tree_flatten(res)
    846   return res_flat

    [... skipping hidden 3 frame]

Cell In[628], line 8, in func(x)
      6 inds = jp.array([0, 2, 4], dtype=jp.int32)
      7 updates = jp.array([100., 101., 102.], dtype=jp.float32)
----> 8 x = x.at[jp.array(inds)].set(updates)
      9 return x

File ~/mambaforge/envs//lib/python3.10/site-packages/jax/_src/numpy/array_methods.py:482, in _IndexUpdateRef.set(self, values, indices_are_sorted, unique_indices, mode)
    473 def set(self, values, *, indices_are_sorted=False, unique_indices=False,
    474         mode=None):
    475   """Pure equivalent of ``x[idx] = y``.
    476 
    477   Returns the value of ``x`` that would result from the NumPy-style
   (...)
    480   See :mod:`jax.ops` for details.
    481   """
--> 482   return scatter._scatter_update(self.array, self.index, values, lax.scatter,
    483                                  indices_are_sorted=indices_are_sorted,
    484                                  unique_indices=unique_indices, mode=mode)

File ~/mambaforge/envs//lib/python3.10/site-packages/jax/_src/ops/scatter.py:78, in _scatter_update(x, idx, y, scatter_op, indices_are_sorted, unique_indices, mode, normalize_indices)
     75 # XLA gathers and scatters are very similar in structure; the scatter logic
     76 # is more or less a transpose of the gather equivalent.
     77 treedef, static_idx, dynamic_idx = jnp._split_index_for_jit(idx, x.shape)
---> 78 return _scatter_impl(x, y, scatter_op, treedef, static_idx, dynamic_idx,
     79                      indices_are_sorted, unique_indices, mode,
     80                      normalize_indices)

File ~/mambaforge/envs//lib/python3.10/site-packages/jax/_src/ops/scatter.py:125, in _scatter_impl(x, y, scatter_op, treedef, static_idx, dynamic_idx, indices_are_sorted, unique_indices, mode, normalize_indices)
    118 # Transpose the gather dimensions into scatter dimensions (cf.
    119 # lax._gather_transpose_rule)
    120 dnums = lax.ScatterDimensionNumbers(
    121   update_window_dims=indexer.dnums.offset_dims,
    122   inserted_window_dims=indexer.dnums.collapsed_slice_dims,
    123   scatter_dims_to_operand_dims=indexer.dnums.start_index_map
    124 )
--> 125 out = scatter_op(
    126   x, indexer.gather_indices, y, dnums,
    127   indices_are_sorted=indexer.indices_are_sorted or indices_are_sorted,
    128   unique_indices=indexer.unique_indices or unique_indices,
    129   mode=mode)
    130 return lax_internal._convert_element_type(out, dtype, weak_type)

    [... skipping hidden 9 frame]

File ~/mambaforge/envs//lib/python3.10/site-packages/jax/experimental/export/shape_poly.py:587, in _DimExpr.__int__(self)
    585   return op.index(next(iter(self._coeffs.values())))
    586 else:
--> 587   raise InconclusiveDimensionOperation(f"Symbolic dimension '{self}' used in a context that requires a constant")

InconclusiveDimensionOperation: Symbolic dimension 'b' used in a context that requires a constant

This error arises for comparison operations with shapes that
are non-constant, and the result of the operation cannot be represented as
a boolean value for all values of the symbolic dimensions involved.
```

### What jax/jaxlib version are you using?

jax v0.4.17, jaxlib v0.4.17, tensorflow v2.14.0

### Which accelerator(s) are you using?

CPU

### Additional system info

Mac

**Comment by gnecula** [comment]
I will look into this, it seems that jvp(scatter) with shape polymorphism is not working properly.

I looked at the code a bit, and there is a simpler code path is the update indices are unique. If you are willing to promise that, then use `set(updates, unique_indices=True)` and I think then it should work.
